### DISCUSSION TITLE
共同开发WebGPU渲染引擎

### DISCUSSION THREAD

**Discussion by yyc-git** [Meta3D-Technology/Meta3D#18] [Ideas]
# 概述

- 只实现渲染
- 基于GPU Driven Pipeline
相关资料可参考：
[GPU Driven Rendering](https://vkguide.dev/docs/gpudriven/gpu_driven_engines/)
[游戏引擎随笔 0x20：UE5 Nanite 源码解析之渲染篇：BVH 与 Cluster 的 Culling](https://zhuanlan.zhihu.com/p/376786001)
[Analysis of UE5 Rendering Technology: Nanite](https://blog.en.uwa4d.com/2022/02/15/analysis-of-ue5-rendering-technology-nanite/)
- 核心功能点
   - 直接光照
   - Frustum Cull
   - Instance Cull
   需要构造BVH
   - Defer Material?
   - 场景数据转为渲染数据（CPU传到GPU）
   - 多线程

# 技术栈

下面语言任选一：
Rescript
TypeScript

# 开发方式

- 每个人独立开发自己的扩展、贡献
- 通过扩展、贡献协议来交互
- 需要有BDD测试覆盖核心代码
- fork项目，提交时发起pr，通过代码审核和ci后merge

# 需要的Demo验证

- 多线程Demo
验证多线程提交Command Buffer;
渲染线程+主线程；
- Draw Indirect
- Frustum Cull
- Instance Cull
需要构造BVH

# 新同学快速上手

请依次阅读下面的资料：

- [Meta3D简介](https://meta3d-website.4everland.app/docs/%E7%AE%80%E4%BB%8B)
- [快速预览](https://meta3d-website.4everland.app/docs/%E5%BF%AB%E9%80%9F%E9%A2%84%E8%A7%88)

请依次完成下面的操作：

- [搭建您的第一个引擎](https://meta3d-website.4everland.app/docs/%E6%90%AD%E5%BB%BA%E6%82%A8%E7%9A%84%E7%AC%AC%E4%B8%80%E4%B8%AA%E5%BC%95%E6%93%8E)
- [搭建您的自定义引擎：实现WebGPU光栅化管线绘制三角形](https://meta3d-website.4everland.app/docs/%E6%90%AD%E5%BB%BA%E6%82%A8%E7%9A%84%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%95%E6%93%8E%EF%BC%9A%E5%AE%9E%E7%8E%B0WebGPU%E5%85%89%E6%A0%85%E5%8C%96%E7%AE%A1%E7%BA%BF%E7%BB%98%E5%88%B6%E4%B8%89%E8%A7%92%E5%BD%A2)

有什么问题请在这里交流，并可加QQ群：561291403

谢谢大家！

**Comment by l1305599284** [comment]
1.引擎的总体设计思路
2.wgsl的拼接思路
3.实现各种材质的渲染效果
4.粒子系统
5.物理系统、刚体碰撞
6.骨骼动画、模型动画
7.物理模拟，如气体扩散，液体流动
